### ISSUE TITLE
Reduced Accuracy with Distributed Training

### ISSUE THREAD

**Issue by hassaankb** [VowpalWabbit/vowpal_wabbit#3763]
I am getting different weights, loss and accuracy when training on the same data with and without spanning server. Is this expected behavior or am I doing something wrong?

#### To Reproduce
1. Run spanning_tree on one node.
2. Using head and tail data from single_machine.sh file.
3. Run commands on another node,
```
cd build/cluster
head -n 100 ../../test/train-sets/0001.dat > head_data
tail -n 100 ../../test/train-sets/0001.dat > tail_data
cat ../../test/train-sets/0001.dat > full_data
```
**Parallel Run**
```
../vowpalwabbit/vw -f native_seed.mdl --random_seed 10 --hash_seed 10 -b 18 -d full_data --cache_file cfull -k --passes 100 --holdout_off &>> /datadrive/logs/logs_vw_native_mdl

../vowpalwabbit/vw -f model1b.mdl --random_seed 10 --hash_seed 10 -b 18 --total 2 --node 1 --unique_id 10 -d tail_data --cache_file c1 -k --passes 100 --span_server 10.244.184.54 --span_server_port 3000 --holdout_off &>> /datadrive/logs/logs_vw_node1_mdl
```
**Default run**
```
../vowpalwabbit/vw -f native_seed.mdl --random_seed 10 --hash_seed 10 -b 18 -d full_data --cache_file cfull -k --passes 100 --holdout_off &>> /datadrive/logs/logs_vw_native_mdl

```
## Output Logs
**Default Run**
```
finished run
number of examples per pass = 200
passes used = 100
weighted example sum = 20000.000000
weighted label sum = 9100.000000
average loss = 0.002087
best constant = 0.455000
best constant's loss = 0.247975
total feature number = 1548200
```
Distributed Run
```
finished run
number of examples per pass = 200
passes used = 100
weighted example sum = 20000.000000
weighted label sum = 9100.000000
average loss = 0.002728
best constant = 0.455000
best constant's loss = 0.247975
total feature number = 1548200

````

#### Environment
Using VW 8.10.2
Ubuntu 18.04.6

**Comment by jackgerrits** [comment]
When using VW in cluster/allreduce/parallel mode degraded learning performance is to be expected. Each learner only gets a subset of the data and weights are averaged to combine the different node results.

When you can train with the entire dataset without parallelization it should be preferred. If parallelizing the least number of nodes is preferred for best learning performance.

**Comment by hassaankb** [comment]
Thanks jackgerrits. Is there any research paper estimating the impact of parallelization/number of nodes on accuracy?

**Comment by jackgerrits** [comment]
You can learn more about allreduce in this paper: https://arxiv.org/abs/1110.4198

**Comment by hassaankb** [comment]
Thanks, I will close this issue.
